bpf-map-pressure-exporter/src/fetcher.go

```diff
@@ -0,0 +1,82 @@
+package main
+
+import (
+	"errors"
+	"os"
+	"strings"
+
+	"github.com/cilium/ebpf"
+	"github.com/cybozu-go/log"
+)
+
+type IBpfMapPressureFetcher interface {
+	Fetch() []bpfMapPressure
+}
+
+type bpfMapPressureFetcher struct {
+	mapNameStrings []string
+}
+
+type bpfMapPressure struct {
+	mapId       uint32
+	mapName     string
+	mapPressure float64
+}
+
+func newFetcher(mapNameStrings []string) *bpfMapPressureFetcher {
+	return &bpfMapPressureFetcher{
+		mapNameStrings: mapNameStrings,
+	}
+}
+
+func (f *bpfMapPressureFetcher) Fetch() []bpfMapPressure {
+	results := []bpfMapPressure{}
+	var id ebpf.MapID = 0
+	var err error
+	for {
+		id, err = ebpf.MapGetNextID(id)
+		if errors.Is(err, os.ErrNotExist) {
+			break
+		}
+		if err != nil {
+			_ = logger.Warn("failed to get next map id", map[string]interface{}{
+				"id":        id,
+				log.FnError: err,
+			})
+			return nil
+		}
+		m, err := ebpf.NewMapFromID(id)
+		if err != nil {
+			_ = logger.Warn("failed to get map", map[string]interface{}{
+				"id":        id,
+				log.FnError: err,
+			})
+			return nil
+		}
+		info, err := m.Info()
+		if err != nil {
+			_ = logger.Warn("failed to get map info", map[string]interface{}{
+				"id":        id,
+				log.FnError: err,
+			})
+			return nil
+		}
+		for _, str := range f.mapNameStrings {
+			if strings.Contains(info.Name, str) {
+				itr := m.Iterate()
```

Comment:
How about using `BatchLookup`.
I'm not sure we can use this for our use case.
We can reduce the number of calling bpf() syscall using this to count elements.
Because `Iterate().Next()` calls bpf() syscall each time...

Please see for details.
- https://pkg.go.dev/github.com/cilium/ebpf#Map.BatchLookup

Reply:
https://github.com/cybozu/neco-containers/pull/1155/commits/f1ac610f8b7f790f8b4143b8240aa79311fb35d5